In [1]:
import pyspark
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .master("local[*]") \
    .appName('homework') \
    .getOrCreate()

Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/02/28 15:40:29 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


### Q1: Install Spark and PySpark

In [2]:
spark.version

'3.2.1'

In [3]:
dataset="fhvhv"

### Q2: HVFHW February 2021

Download the HVFHV data for february 2021:

```bash
wget https://nyc-tlc.s3.amazonaws.com/trip+data/fhvhv_tripdata_2021-02.csv
```

Read it with Spark using the same schema as we did 
in the lessons. We will use this dataset for all
the remaining questions.

Repartition it to 24 partitions and save it to
parquet.

What's the size of the folder with results (in MB)?

In [20]:
!mkdir data/raw
!mkdir data/raw/fhvhv

mkdir: data/raw: File exists
mkdir: data/raw/fhvhv: File exists


In [25]:
!wget  -O ./data/raw/fhvhv/fhvhv_tripdata_2021-02.csv https://nyc-tlc.s3.amazonaws.com/trip+data/fhvhv_tripdata_2021-02.csv

--2022-02-28 15:46:33--  https://nyc-tlc.s3.amazonaws.com/trip+data/fhvhv_tripdata_2021-02.csv
Resolving nyc-tlc.s3.amazonaws.com (nyc-tlc.s3.amazonaws.com)... 52.217.160.9
Connecting to nyc-tlc.s3.amazonaws.com (nyc-tlc.s3.amazonaws.com)|52.217.160.9|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 733822658 (700M) [text/csv]
Saving to: ‘./data/raw/fhvhv/fhvhv_tripdata_2021-02.csv’

./data/raw/fhvhv/fh 100%[===================>] 699,83M  10,2MB/s    in 88s     

2022-02-28 15:48:01 (8,00 MB/s) - ‘./data/raw/fhvhv/fhvhv_tripdata_2021-02.csv’ saved [733822658/733822658]



In [26]:
from pyspark.sql import types
schema = types.StructType([
    types.StructField("hvfhs_license_num", types.StringType(), True),
    types.StructField("dispatching_base_num", types.StringType(), True),
    types.StructField("pickup_datetime", types.TimestampType(), True),
    types.StructField("dropoff_datetime", types.TimestampType(), True),
    types.StructField("PULocationID", types.IntegerType(), True),
    types.StructField("DOLocationID", types.IntegerType(), True),
    types.StructField("SR_Flag", types.StringType(), True)
])

In [27]:
df = spark.read \
    .option("header", "true") \
    .schema(schema)\
    .csv('data/raw/fhvhv/fhvhv_tripdata_2021-02.csv')

In [28]:
df

DataFrame[hvfhs_license_num: string, dispatching_base_num: string, pickup_datetime: timestamp, dropoff_datetime: timestamp, PULocationID: int, DOLocationID: int, SR_Flag: string]

In [30]:
df.repartition(24) \
    .write.parquet('data/pq/fhvhv/', mode="overwrite")

22/02/28 15:53:23 WARN MemoryManager: Total allocation exceeds 95,00% (1.020.054.720 bytes) of heap memory
Scaling row group sizes to 95,00% for 8 writers
22/02/28 15:53:23 WARN MemoryManager: Total allocation exceeds 95,00% (1.020.054.720 bytes) of heap memory
Scaling row group sizes to 84,44% for 9 writers
22/02/28 15:53:23 WARN MemoryManager: Total allocation exceeds 95,00% (1.020.054.720 bytes) of heap memory
Scaling row group sizes to 76,00% for 10 writers
22/02/28 15:53:23 WARN MemoryManager: Total allocation exceeds 95,00% (1.020.054.720 bytes) of heap memory
Scaling row group sizes to 69,09% for 11 writers
22/02/28 15:53:23 WARN MemoryManager: Total allocation exceeds 95,00% (1.020.054.720 bytes) of heap memory
Scaling row group sizes to 63,33% for 12 writers
22/02/28 15:53:28 WARN MemoryManager: Total allocation exceeds 95,00% (1.020.054.720 bytes) of heap memory
Scaling row group sizes to 69,09% for 11 writers
22/02/28 15:53:28 WARN MemoryManager: Total allocation exceeds 95,

In [31]:
# list the files in folder
! ls data/pq/fhvhv/ 
! du -sh data/pq/fhvhv/

_SUCCESS
part-00000-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00001-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00002-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00003-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00004-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00005-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00006-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00007-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00008-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00009-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00010-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00011-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00012-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00013-bfbd0825-dfa6-4a4b-9fc8-db889928f38c-c000.snappy.parquet
part-00014-bfbd0825-dfa6-4a4b-9fc8-db88

**Answer:** 221M -> 208MB

### Prep Steps to run SQL queries on spark dataframe

In [32]:
df_fhvhv = spark.read.parquet('data/pq/fhvhv/*')

In [33]:
df_fhvhv.registerTempTable('fhvhv')

/Users/iremertuerk/workspace-personal/Data-Engineering-Zoomcamp-All/dt-de-zoomcamp/venv/lib/python3.9/site-packages/pyspark/sql/dataframe.py:138: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  warnings.warn(


In [34]:
df_fhvhv.show()

+-----------------+--------------------+-------------------+-------------------+------------+------------+-------+
|hvfhs_license_num|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|
+-----------------+--------------------+-------------------+-------------------+------------+------------+-------+
|           HV0003|              B02864|2021-02-04 02:19:32|2021-02-04 02:30:59|          20|         220|   null|
|           HV0003|              B02887|2021-02-02 14:23:12|2021-02-02 14:44:21|          61|          95|   null|
|           HV0005|              B02510|2021-02-01 01:32:24|2021-02-01 01:54:28|          41|         186|   null|
|           HV0005|              B02510|2021-02-03 13:17:38|2021-02-03 13:24:43|         198|         160|   null|
|           HV0005|              B02510|2021-02-03 08:17:43|2021-02-03 08:43:33|         140|          68|   null|
|           HV0003|              B02864|2021-02-02 19:01:04|2021-02-02 19:07:22|

### Q3: Count records 

How many taxi trips were there on February 15?

Consider only trips that started on February 15.

In [35]:
q3 = """
SELECT
    date_trunc('day', pickup_datetime) AS day, 
    COUNT(*)
FROM
    fhvhv
WHERE
    date_trunc('day', pickup_datetime) == '2021-02-15'
GROUP BY 1
"""

df_Q3 = spark.sql(q3)

In [36]:
df_Q3.show()

+-------------------+--------+
|                day|count(1)|
+-------------------+--------+
|2021-02-15 00:00:00|  367170|
+-------------------+--------+



**Answer:** 367170

### Q4: Longest trip for each day

Now calculate the duration for each trip.

Trip starting on which day was the longest? 

In [37]:
from pyspark.sql.functions import col, pandas_udf

def duration(pickup, dropoff):
    return (dropoff-pickup)

duration_udf = pandas_udf(duration, returnType= types.LongType())

In [38]:
df_Q4 = df_fhvhv.withColumn("duration", duration_udf(col("pickup_datetime"), col("dropoff_datetime")))

In [39]:
df_Q4.registerTempTable('fhvhv_with_duration')

/Users/iremertuerk/workspace-personal/Data-Engineering-Zoomcamp-All/dt-de-zoomcamp/venv/lib/python3.9/site-packages/pyspark/sql/dataframe.py:138: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  warnings.warn(


In [40]:
q4 = """
SELECT
    pickup_datetime,
    duration
FROM
    fhvhv_with_duration
ORDER BY duration DESC
"""

df_Q4 = spark.sql(q4)

In [41]:
df_Q4.show()

+-------------------+--------------+
|    pickup_datetime|      duration|
+-------------------+--------------+
|2021-02-11 13:40:44|75540000000000|
|2021-02-17 15:54:53|57221000000000|
|2021-02-20 12:08:15|44039000000000|
|2021-02-03 20:24:25|40653000000000|
|2021-02-19 23:17:44|37577000000000|
|2021-02-25 17:13:35|35010000000000|
|2021-02-20 01:36:13|34806000000000|
|2021-02-18 15:24:19|34612000000000|
|2021-02-18 01:31:20|34555000000000|
|2021-02-10 20:51:39|34169000000000|
|2021-02-10 01:56:17|32476000000000|
|2021-02-25 09:18:18|32439000000000|
|2021-02-21 19:59:13|32223000000000|
|2021-02-09 18:36:13|32087000000000|
|2021-02-06 09:48:09|31447000000000|
|2021-02-02 09:42:30|30913000000000|
|2021-02-10 10:12:08|30856000000000|
|2021-02-09 13:30:13|30732000000000|
|2021-02-21 22:50:52|30660000000000|
|2021-02-05 21:32:33|30511000000000|
+-------------------+--------------+
only showing top 20 rows



**Answer:** "2021-02-11"

### Q5: Most frequent `dispatching_base_num`

Now find the most frequently occurring `dispatching_base_num` 
in this dataset.

How many stages this spark job has?

> Note: the answer may depend on how you write the query,
> so there are multiple correct answers. 
> Select the one you have.

In [42]:
q5 = """
SELECT
    dispatching_base_num,
    COUNT(*) AS occurence
FROM
    fhvhv
GROUP BY 1
ORDER BY 2 DESC
"""

df_Q5 = spark.sql(q5)

In [43]:
df_Q5.show()

+--------------------+---------+
|dispatching_base_num|occurence|
+--------------------+---------+
|              B02510|  3233664|
|              B02764|   965568|
|              B02872|   882689|
|              B02875|   685390|
|              B02765|   559768|
|              B02869|   429720|
|              B02887|   322331|
|              B02871|   312364|
|              B02864|   311603|
|              B02866|   311089|
|              B02878|   305185|
|              B02682|   303255|
|              B02617|   274510|
|              B02883|   251617|
|              B02884|   244963|
|              B02882|   232173|
|              B02876|   215693|
|              B02879|   210137|
|              B02867|   200530|
|              B02877|   198938|
+--------------------+---------+
only showing top 20 rows



**Answer:** 2 

### Question 6: Most common locations pair

Find the most common pickup-dropoff pair. 

For example:

"Jamaica Bay / Clinton East"

Enter two zone names separated by a slash

If any of the zone names are unknown (missing), use "Unknown". For example, "Unknown / Clinton East". 

#### Step 0: Prep Data

In [46]:
!mkdir data/raw/zones

In [47]:
!wget -O ./data/raw/zones/taxi+_zone_lookup.csv https://s3.amazonaws.com/nyc-tlc/misc/taxi+_zone_lookup.csv

--2022-02-28 15:57:17--  https://s3.amazonaws.com/nyc-tlc/misc/taxi+_zone_lookup.csv
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.217.128.72
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.217.128.72|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12322 (12K) [application/octet-stream]
Saving to: ‘./data/raw/zones/taxi+_zone_lookup.csv’

./data/raw/zones/ta 100%[===================>]  12,03K  --.-KB/s    in 0,002s  

2022-02-28 15:57:17 (7,17 MB/s) - ‘./data/raw/zones/taxi+_zone_lookup.csv’ saved [12322/12322]



In [49]:
df_zones = spark.read \
    .option("header", "true") \
    .csv('data/raw/zones/taxi+_zone_lookup.csv')

In [51]:
df_zones.write.parquet('data/pq/zones', mode="overwrite")

In [52]:
df_zones = spark.read.parquet('data/pq/zones/')

In [68]:
df_fhvhv.registerTempTable('fhvhv')
df_zones.registerTempTable('zones')

#### Step1: Join the Zone data with trip data

In [69]:
#     fhvhv.pickup_datetime,
#     fhvhv.dropoff_datetime,
#     fhvhv.PULocationID,
#     fhvhv.DOLocationID,
#     fhvhv.SR_Flag,

q6_1 = """
SELECT
    fhvhv.hvfhs_license_num,
    fhvhv.dispatching_base_num,    
    pickup_zone.Zone AS pickup_zone,
    dropoff_zone.Zone AS dropoff_zone
FROM fhvhv
    LEFT JOIN zones AS pickup_zone
    on fhvhv.PULocationID = pickup_zone.locationid
    LEFT JOIN zones AS dropoff_zone
    on fhvhv.DOLocationID = dropoff_zone.locationid
"""
df_Q6 = spark.sql(q6_1)

In [70]:
import pyspark.sql.functions as F
from pyspark.sql.functions import coalesce


# df_Q6 = df_Q6 \
#     .withColumn("pickup_zone",coalesce(df_Q6.pickup_zone, F.lit('Unkown'))) \
#     .withColumn("dropoff_zone",coalesce(df_Q6.dropoff_zone, F.lit('Unkown'))) 

df_Q6 = df_Q6 \
    .na.fill("Unknown","pickup_zone") \
    .na.fill("Unknown","dropoff_zone") \

df_Q6 = df_Q6.withColumn('loc_pairs', F.concat_ws(' / ', F.col('pickup_zone'), F.col('dropoff_zone')))
# df_Q6 = df_Q6.withColumn('loc_pairs', F.concat(F.col('pickup_zone'), F.lit(' / '), F.col('dropoff_zone')))


In [71]:
df_Q6.registerTempTable('df_Q6')

In [72]:
q6_2 = """
SELECT
    loc_pairs,
    COUNT(*) as occurence
FROM df_Q6
GROUP BY loc_pairs
ORDER BY 2 DESC
"""
df_Q6 = spark.sql(q6_2)

In [73]:
df_Q6.show()

+--------------------+---------+
|           loc_pairs|occurence|
+--------------------+---------+
|East New York / E...|    45041|
|Borough Park / Bo...|    37329|
| Canarsie / Canarsie|    28026|
|Crown Heights Nor...|    25976|
|Bay Ridge / Bay R...|    17934|
|Jackson Heights /...|    14688|
|   Astoria / Astoria|    14688|
|Central Harlem No...|    14481|
|Bushwick South / ...|    14424|
|Flatbush/Ditmas P...|    13976|
|South Ozone Park ...|    13716|
|Brownsville / Bro...|    12829|
|    JFK Airport / NA|    12542|
|Prospect-Lefferts...|    11814|
|Forest Hills / Fo...|    11548|
|Bushwick North / ...|    11491|
|Bushwick South / ...|    11487|
|Crown Heights Nor...|    11462|
|Crown Heights Nor...|    11342|
|Prospect-Lefferts...|    11308|
+--------------------+---------+
only showing top 20 rows



In [74]:
df_Q6.head(1)

[Row(loc_pairs='East New York / East New York', occurence=45041)]

**Answer**: [Row(loc_pairs='East New York / East New York', occurence=45041)]

### Bonus question. Join type

(not graded) 

For finding the answer to Q6, you'll need to perform a join.

What type of join is it?

And how many stages your spark job has?

answer:

- As we are also interested to list Unknown zone names as well, we can not use INNER JOIN
- On the other hand, we dont care about the zones that are not exists in our trip data, therefore FULL OUTER JOIN would create more rows than we actually need
- Our main focus is to keep all tripdata therefore LEFT JOIN should be used.